### PR TITLE
implement aten.conv1d, aten.conv3d, and aten.conv_tbc

### PR DESCRIPTION
**Pull request by renxida**
convolution with [time,batch,channel] ordering, as opposed to the default [batch, channel, time]. Currently implementing by transposing the input and output, but may need to get its own implementation in the future because this is supposed to be an op that gives a speedup. This is used by fairseq (https://github.com/facebookresearch/fairseq/issues/172).

(in case you were wondering like me, this is different from transposed convolution. Transposed convolution has fractional strides).

### CODE REVIEW

**Comment by renxida** [comment]
Currently getting some shape inference related trouble: [my conv_tbc_smoketest.py](https://gist.github.com/renxida/3c8477a9181318c4a8c43ed6ef3bd59b) gives the following error:

<details>
<summary> unsupported by backend contract: tensor with unknown rank </summar>
```
/home/azureuser/miniconda/lib/python3.11/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
TEST: test_enable_ir_printing
Traceback (most recent call last):
  File "/home/azureuser/torch-mlir/./scratch/conv_tbc_example.py", line 34, in <module>
    @run_test
     ^^^^^^^^
  File "/home/azureuser/torch-mlir/./scratch/conv_tbc_example.py", line 15, in run_test
    f()
  File "/home/azureuser/torch-mlir/./scratch/conv_tbc_example.py", line 40, in test_enable_ir_printing
    torch_mlir.compile(TinyModel(),
  File "/home/azureuser/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/__init__.py", line 458, in compile
    run_pipeline_with_repro_report(
  File "/home/azureuser/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir/compiler_utils.py", line 73, in run_pipeline_with_repro_report
    raise TorchMlirCompilerError(trimmed_message) from None
torch_mlir.compiler_utils.TorchMlirCompilerError: Lowering TorchScript IR -> Torch Backend IR failed with the following diagnostics:


python exception: Failure while executing pass pipeline:
error: "aten::conv_tbc"("/home/azureuser/torch-mlir/./scratch/conv_tbc_example.py":27:22): unsupported by backend contract: tensor with unknown rank
note: "aten::conv_tbc"("/home/azureuser/torch-mlir/./scratch/conv_tbc_example.py":27:22): see current operation: %1 = "torch.aten.conv_tbc"(%arg0, %arg1, %arg2, %0) : (!torch.vtensor<[3,1,5],f32>, !torch.vtensor<[5,15],f32>, !torch.vtensor<[5],f32>, !torch.int) -> !torch.vtensor<*,f32>
note: "aten::conv_tbc"("/home/azureuser/torch-mlir/./scratch/conv_tbc_example.py":27:22): this is likely due to a missing transfer function in abstract_interp_lib_gen.py

For Torch-MLIR developers, the error can be reproduced with:
$ torch-mlir-opt -pass-pipeline='builtin.module(torchscript-module-to-torch-backend-pipeline{backend-legal-ops= extra-library=})' /tmp/TinyModel.mlir
Add '-mlir-print-ir-after-all -mlir-disable-threading' to get the IR dump for debugging purpose.
```

</details>

File: projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py [hunk 1]
```diff
@@ -1127,6 +1127,13 @@ def aten〇conv2d〡shape(input: List[int], weight: List[int], bias: Optional[Li
 def aten〇conv_transpose2d〇input〡shape(input: List[int], weight: List[int], bias: Optional[List[int]] = None, stride: List[int] = (1, 1,), padding: List[int] = (0, 0,), output_padding: List[int] = (0, 0,), groups: int = 1, dilation: List[int] = (1, 1,)) -> List[int]:
     return upstream_shape_functions.conv_transpose2d_input(input, weight, bias, stride, padding, output_padding, groups, dilation)
 
+def aten〇conv_tbc〡shape(self: List[int], weight: List[int], bias: List[int], pad: int = 0) -> List[int]:
```
Comment (wu-s-john) [inline projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py:1130]: You should make a test of this as mentioned here.

https://github.com/llvm/torch-mlir/pull/2735/files#r1453668408

Comment (renxida) [reply projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py:1130]: ty & willdo!

**Comment by renxida** [comment]
```
    def test_conv_tbc(self):
        with set_default_dtype(torch.double):
            inp = torch.randn(9, 4, 5, requires_grad=True)
            weight = torch.randn(3, 5, 6, requires_grad=True)
            bias = torch.randn(6, requires_grad=True)

            gradcheck(lambda i, w, b, pad: F.conv_tbc(i, w, b, pad), (inp, weight, bias, 3))
```

example shapes for conv_tbc from https://github.com/pytorch/pytorch/blob/3e8c8ce37bbfaafa8581fb48506c0a70ea54463d/test/nn/test_convolution.py#L623

**Comment by renxida** [comment]
Shape is good now. Realized that we currently don't have conv1d implemented, so I'm going to go ahead and do that.
Todo list:
- write conv1d tests
- write conv1d

**Comment by renxida** [comment]
This will also resolve #1879

**Comment by vivekkhandelwal1** [comment]
@renxida, I think I mentioned this in your last PR as well, that we should get atleast one approval before merging the PR.

**Comment by stellaraccident** [comment]
Thank you for flagging, Vivek. Agreed.

While we do sometimes allow for post submit reviews, that is typically for urgent project maintenance or trivial/non functional changes -- neither of which applies here.

Usually on the main project, I would revert a patch like this. @vivekkhandelwal1 your call: if you review and think it is ok, we can keep it. But if there are comments or updates needed, please revert.

**Comment by hanhanW** [comment]
FYI, this PR introduced unused variables which could break downstream projects with `-Wunused-variable` build.

```
/work/third_party/torch-mlir/lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2620:13: error: unused variable 'zeroList' [-Werror,-Wunused-variable]
      Value zeroList = rewriter.create<PrimListConstructOp>(
```

Sample log: https://github.com/openxla/iree/actions/runs/7686712952/job/20945886285

**Comment by stellaraccident** [comment]
> FYI, this PR introduced unused variables which could break downstream projects with `-Wunused-variable` build.
> 
> ```
> /work/third_party/torch-mlir/lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:2620:13: error: unused variable 'zeroList' [-Werror,-Wunused-variable]
>       Value zeroList = rewriter.create<PrimListConstructOp>(
> ```
> 
> Sample log: https://github.com/openxla/iree/actions/runs/7686712952/job/20945886285

We should tighten the warning flags on this project.

**Comment by hanhanW** [comment]
> We should tighten the warning flags on this project.

I'm +1 on this. This is second time that I hit the issue when I'm doing integration. If the CI on this project could catch it, that's would be very helpful.

**Comment by vivekkhandelwal1** [comment]
I'm reverting this PR, since the PR doesn't follow the contribution standards and guidelines. @renxida, after I revert this PR, please create a seperate new PR and I will add the review comments in that PR.

**Comment by vivekkhandelwal1** [comment]
> I'm reverting this PR, since the PR doesn't follow the contribution standards and guidelines. @renxida, after I revert this PR, please create a seperate new PR and I will add the review comments in that PR.

@stellaraccident, I tried reverting this PR but it's giving some error. What should I do?

**Comment by stellaraccident** [comment]
> > I'm reverting this PR, since the PR doesn't follow the contribution standards and guidelines. @renxida, after I revert this PR, please create a seperate new PR and I will add the review comments in that PR.
> 
> @stellaraccident, I tried reverting this PR but it's giving some error. What should I do?

It is easier to help if error messages are given. I loosened the branch protection. Try now.

**Comment by vivekkhandelwal1** [comment]
> > > I'm reverting this PR, since the PR doesn't follow the contribution standards and guidelines. @renxida, after I revert this PR, please create a seperate new PR and I will add the review comments in that PR.
> > 
> > 
> > @stellaraccident, I tried reverting this PR but it's giving some error. What should I do?
> 
> It is easier to help if error messages are given. I loosened the branch protection. Try now.

I'm still not able to do so. Here's the error message:
```
Sorry, this pull request couldn’t be reverted automatically. It may have already been reverted, or the content may have changed since it was merged.
```
Sorry for not sharing it before.

**Pull request merged**